File: clutter.py
from flask import Flask, render_template, request
from html import escape
from pathlib import Path
from urllib.parse import quote
import collections
import datetime
import difflib
import json
import logging
import os
import re
import shutil
import sys

logging.getLogger('werkzeug').disabled = True
path = None

def read():
    with open(path, 'r') as f:
        lines = f.read().split('\n')
    return lines, len(lines)

def write(lines):
    with open(path, 'w') as f:
        f.write('\n'.join(lines))

def change(old, new):
    timestamp = datetime.datetime.now().strftime(f'%Y%m%d%H%M%S')
    Path(log_path).touch()
    with open(log_path, 'r') as f:
        original = f.read()
    if old == '':
        delta = new
    elif new == '':
        delta = f'{timestamp}\n{old}'
    else:
        delta = f'{timestamp}\n{new}\n{old}'
    with open(log_path, 'w') as f:
        f.write(delta + '\n\n' + original)

def push(content):
    header = datetime.datetime.now().strftime('%Y%m%d%H%M%S ')
    change('', header + content)
    write([header + content] + read()[0])

def parse(q):
    query = {'q': q, 'units': [], 'sort': None, 'replace': None}
    for word in q.split():
        head, tail = word[0], word[1:]
        convert = lambda word: word.replace('[space]', ' ').lower()
        if head == 'S' and len(word) >= 4 and word[1] == word[-1]:
            chunks = word.split(word[1])
            if len(chunks) != 4:
                continue
            old = chunks[1].replace('[space]', ' ')
            new = chunks[2].replace('[space]', ' ')
            query['replace'] = {'old': old, 'new': new}
            query['units'].append({'match': True, 'word': old.lower()})
        elif len(word) > 1 and head in {'A', 'D'}:
            query['sort'] = {'reverse': head == 'D', 'by': convert(tail)}
            query['units'].append({'match': True, 'word': convert(tail)})
        elif len(word) > 1 and head == 'N':
            query['units'].append({'match': False, 'word': convert(tail)})
        else:
            query['units'].append({'match': True, 'word': convert(word)})
    return query

def process(query):
    (lines, n), data, items = read(), {}, []
    units, sort, replace = query['units'], query['sort'], query['replace']
    for i, line in enumerate(lines):
        if line == '':
            continue
        line_ = line.lower()
        if all(unit['match'] == (unit['word'] in line_) for unit in units):
            items.append({'id': n - i, 'line': line})
    if sort:
        key = lambda item: item['line'].lower().split(sort['by'], 1)[1]
        items = sorted(items, key=key, reverse=sort['reverse'])
    if replace:
        check_case = lambda item: replace['old'] in item['line']
        items = list(filter(check_case, items))
        old, new = escape(replace['old']), escape(replace['new'])
        q = quote(replace['new'].lower().replace(' ', '[space]'))
        link = f'<a href="/?q={q}">{new}</a>'
        warning = f'"{link}" already exists!'
        data['message'] = warning if new in '\n'.join(lines) else ''
        for item in items:
            line = escape(item['line'])
            item['href'] = '/?q=' + quote(item['line'][:14])
            item['line'] = line.replace(old, f'<del>{old}</del>')
            item['revision'] = line.replace(old, f'<ins>{new}</ins>')
    else:
        date = lambda k: items[k]['line'][:8]
        for i in range(len(items)):
            if sort:
                items[i]['left'] = '─'
                continue
            begin = (i == 0 or items[i - 1]['left'] in {'┴', '─'})
            close = (i == len(items) - 1 or date(i + 1) != date(i))
            items[i]['left'] = [['┼', '┴'], ['┬', '─']][begin][close]
    count = '&nbsp;' * (14 - len(str(len(items)))) + str(len(items)) + '&nbsp;'
    data['title'], data['q'] = query['q'] or 'Clutter', query['q']
    data['items'], data['count'] = items, count
    return data

app = Flask(__name__)

@app.route('/')
def index():
    request.args.get('add') and push(request.args.get('add'))
    q = request.args.get('q') or ''
    query = parse(q)
    data = process(query)
    if query['replace']:
        return render_template('replace.html', data=data)
    else:
        data['clock'] = datetime.datetime.now().strftime(f'%Y%m%d%H%M%S')
        return render_template('index.html', data=data)

@app.route('/tag')
def tag():
    request.args.get('add') and push(request.args.get('add'))
    q = request.args.get('q') or ''
    query = parse(q)
    data = process(query)
    tags = collections.Counter()
    for item in data['items']:
        for word in item['line'].split(' '):
            if word.startswith('#'):
                tags[word] += 1
    data['title'] = data['q'] or 'Tag'
    data['tags'] = tags.most_common()
    return render_template('tag.html', data=data)

@app.route('/log')
def log():
    q = request.args.get('q') or ''
    words = [word.replace('[space]', ' ').lower() for word in q.split()]
    with open(log_path, 'r') as f:
        blocks = f.read().split('\n\n')
    items = []
    total = 0
    tokenize = lambda line: list(filter(None, re.split(r'(\W)', line)))
    for block in blocks:
        block_ = block.lower()
        if not all(word in block_ for word in words):
            continue
        total += 1
        if total > 1000:
            continue
        complete = False
        lines = block.split('\n')
        if len(lines) == 1:
            timestamp = lines[0][:14]
        elif len(lines) == 2:
            timestamp = lines[1][:14]
            lines[1] = f'<del>{escape(lines[1])}</del>'
        elif len(lines) == 3:
            timestamp = lines[1][:14]
            if '#plan' in lines[1] and '|' in lines[1]:
                complete = lines[0][:8] == lines[1][:8]
            old, new = '', ''
            for diff in difflib.ndiff(tokenize(lines[2]), tokenize(lines[1])):
                code, token = diff[0], escape(diff[2:])
                if code == ' ':
                    old += token
                    new += token
                elif code == '-':
                    old += f'<del>{token}</del>'
                elif code == '+':
                    new += f'<ins>{token}</ins>'
            lines[1], lines[2] = new, old
        else:
            continue
        yyyy = f'<span class="timestamp">{escape(lines[0][:4])}</span>'
        log_q = f'/log?q={quote(lines[0][:8])}#{quote(lines[0][:14])}'
        mmdd = f'<a href="{log_q}">{escape(lines[0][4:8])}</a>'
        hhmmss = f'<span class="timestamp">{escape(lines[0][8:14])}</span>'
        body_ = escape(lines[0][14:])
        body = f'<ins>{body_}</ins>' if body_ else ''
        lines[0] = yyyy + mmdd + hhmmss + body
        href = '/?q=' + quote(timestamp)
        items.append({'href': href, 'lines': lines, 'complete': complete})
    data = {'title': q or 'Log', 'q': q, 'items': items, 'total': total}
    return render_template('log.html', data=data)

@app.route('/add', methods=['POST'])
def add():
    push(request.data.decode('utf-8'))
    return 'ok'

@app.route('/edit', methods=['POST'])
def edit():
    item = json.loads(request.data.decode('utf-8'))
    lines, n = read()
    index = n - int(item['id'])
    if lines[index] != item['line']:
        return 'no'
    change(lines[index], item['revision'])
    lines[index] = item['revision']
    write(lines)
    return 'ok'

@app.route('/replace', methods=['POST'])
def replace():
    backup()
    items = json.loads(request.data.decode('utf-8'))
    lines, n = read()
    deltas = []
    for item in items:
        index = n - int(item['id'])
        if lines[index] != item['line']:
            return 'no'
        deltas.append((lines[index], item['revision']))
        lines[index] = item['revision']
    for delta in deltas:
        change(delta[0], delta[1])
    write(lines)
    return 'ok'

def backup():
    snapshots = os.path.join(os.path.dirname(path), 'snapshots')
    if not os.path.exists(snapshots):
        os.mkdir(snapshots)
    name = datetime.datetime.now().strftime('%Y%m%d%H%M%S.clutter.txt')
    shutil.copyfile(path, os.path.join(snapshots, name))
    return 'ok'

@app.template_filter('link')
def link(line):
    parts = []
    tag = {'#': 'hash', '$': 'dollar'}
    for word in line.split(' '):
        if word.startswith('http'):
            parts.append(f'<a href="{escape(word)}">{escape(word)}</a>')
        elif word and word[0] in tag:
            parts.append(f'<span class="{tag[word[0]]}">{escape(word)}</span>')
        elif word.startswith('/') and len(word) > 1:
            parts.append(f'<a href="file://{escape(word)}">{escape(word)}</a>')
        else:
            parts.append(escape(word))
    line = ' '.join(parts)
    if re.match(r'\d{14}', line):
        yyyy = f'<span class="timestamp">{line[:4]}</span>'
        mmdd = f'<a href="/?q={line[:8]}#{line[:14]}">{line[4:8]}</a>'
        hh = f'<span class="timestamp">{line[8:10]}</span>'
        mmss = f'<a href="/log?q={line[:14]}">{line[10:14]}</a>'
        body = line[14:]
        line = yyyy + mmdd + hh + mmss + body
    return line

if __name__ == '__main__':
    argv = sys.argv
    if len(argv) < 2:
        print('Path required.')
        sys.exit()
    path = argv[1]
    log_path = path + '.log'
    port = argv[2] if len(argv) > 2 else '12224'
    app.run(port=port)

File: clutter.js
if (!localStorage.clutter) {
    localStorage.clutter = '';
}

const escape = x => {
    x = x.replaceAll('&', '&amp;');
    x = x.replaceAll('"', '&quot;');
    x = x.replaceAll("'", '&apos;');
    x = x.replaceAll('<', '&lt;');
    x = x.replaceAll('>', '&gt;');
    return x;
};

const read = () => {
    const lines = localStorage.clutter.split('\n');
    return [lines, lines.length];
};

const write = lines => {
    localStorage.clutter = lines.join('\n');
};

const header = () => {
    const time = new Date();
    const year = time.getFullYear();
    const month = format(time.getMonth() + 1);
    const date = format(time.getDate());
    const hour = format(time.getHours());
    const minute = format(time.getMinutes());
    const second = format(time.getSeconds());
    return year + month + date + hour + minute + second + ' ';
};

const push = content => {
    write([header() + content].concat(read()[0]));
};

const parse = q => {
    const query = { q: q, units: [], sort: null, swap: null };
    for (const word of q.trim().split(/\s+/)) {
        const head = word[0];
        const tail = word.slice(1);
        const convert = word => word.replaceAll('[space]', ' ').toLowerCase();
        if (head === 'S' && word.length >= 4 && word[1] === word.slice(-1)) {
            const chunks = word.split(word[1]);
            if (chunks.length !== 4) { continue; }
            const source = chunks[1].replaceAll('[space]', ' ');
            const target = chunks[2].replaceAll('[space]', ' ');
            query.swap = { source: source, target: target };
            query.units.push({ match: true, word: source.toLowerCase() });
        }
        else if (word.length > 1 && ['A', 'D'].includes(head)) {
            query.sort = { reverse: head === 'D', by: convert(tail) };
            query.units.push({ match: true, word: convert(tail) });
        }
        else if (word.length > 1 && head === 'N') {
            query.units.push({ match: false, word: convert(tail) });
        }
        else {
            query.units.push({ match: true, word: convert(word) });
        }
    }
    return query;
};

const process = query => {
    const [lines, n] = read();
    const data = {};
    let items = [];
    const { units, sort, swap } = query;
    for (const [i, line] of lines.entries()) {
        if (line === '') {
            continue
        }
        const line_ = line.toLowerCase();
        if (units.every(unit => unit.match === line_.includes(unit.word))) {
            items.push({ id: n - i, line: line });
        }
    }
    if (sort) {
        const key = item => {
            const lower = item.line.toLowerCase();
            return lower.slice(lower.indexOf(sort.by));
        };
        const compare = (a, b) => {
            if (key(a) > key(b)) { return 1; }
            else if (key(a) == key(b)) { return 0; }
            else { return -1; }
        };
        const compareFn = (a, b) => {
            return sort.reverse ? compare(b, a) : compare(a, b);
        };
        items = items.sort(compareFn);
    }
    if (swap) {
        const checkCase = item => item.line.includes(swap.source);
        items = items.filter(checkCase);
        const { source, target } = swap;
        const lower = target.toLowerCase();
        const q = encodeURIComponent(lower.replaceAll(' ', '[space]'));
        const link = `<a href="?q=${q}">${escape(target)}</a>`;
        const warning = `"${link}" already exists!`;
        data.message = lines.join('\n').includes(target) ? warning : '';
        for (const item of items) {
            item.revision = item.line.replaceAll(source, target);
        }
    }
    const date = k => items[k].line.slice(0, 8);
    for (let i = 0; i < items.length; i++) {
        const begin = +(i === 0 || ['┴', '─'].includes(items[i - 1].left));
        const close = +(i == items.length - 1 || date(i + 1) !== date(i));
        items[i].left = [['┼', '┴'], ['┬', '─']][begin][close];
    }
    const length = items.length.toString();
    const count = '&nbsp;'.repeat(14 - length.length) + length + '&nbsp;';
    data.title = query.q || 'Clutter';
    data.q = query.q;
    data.items = items;
    data.count = count;
    return data;
};

const createSwap = item => {
    const row = document.createElement('div');
    row.id = item.id;
    document.getElementById('swapRows').appendChild(row);
    const originalRow = document.createElement('div');
    originalRow.className = 'row';
    row.appendChild(originalRow);
    const originalLeft = document.createElement('div');
    originalLeft.className = 'left';
    originalLeft.innerText = '-';
    originalRow.appendChild(originalLeft);
    const originalLine = document.createElement('pre');
    originalLine.className = 'original';
    originalLine.innerHTML = item.line;
    originalRow.appendChild(originalLine);
    const revisionRow = document.createElement('div');
    revisionRow.className = 'row';
    row.appendChild(revisionRow);
    const revisionLeft = document.createElement('div');
    revisionLeft.className = 'left';
    revisionLeft.innerText = '+';
    revisionRow.appendChild(revisionLeft);
    const revisionLine = document.createElement('pre');
    revisionLine.className = 'revision';
    revisionLine.innerHTML = item.revision;
    revisionRow.appendChild(revisionLine);
    const stuff = document.createElement('div');
    stuff.className = 'stuff';
    row.appendChild(stuff);
};

const onSwap = data => {
    document.getElementById('gap').style.display = 'block';
    document.getElementById('space').style.display = 'block';
    document.getElementById('swapRows').style.display = 'block';
    const button = '&nbsp;'.repeat(5) + 'swap' + '&nbsp;'.repeat(5);
    document.getElementById('swap').innerHTML = button;
    document.getElementById('message').innerHTML = '&nbsp;' + data.message;
    for (const item of data.items) {
        createSwap(item);
    }
    const swap = document.getElementById('swap');
    swap.addEventListener('click', async event => {
        if (!confirm('Are you sure?')) { return; }
        if (!postSwap(data.items)) { alert('Error!'); }
        location.href = location.pathname;
    });
};

const createIndex = item => {
    const row = document.createElement('div');
    row.className = 'row';
    row.id = item.id;
    document.getElementById('indexRows').appendChild(row);
    const edit = document.createElement('div');
    edit.className = 'btn left';
    edit.innerText = item.left;
    row.appendChild(edit);
    const pre = document.createElement('pre');
    pre.innerHTML = link(item.line);
    row.appendChild(pre);
    const none = document.createElement('div');
    none.className = 'left none';
    none.innerText = '*';
    row.appendChild(none);
    const input = document.createElement('input');
    input.className = 'none';
    input.spellcheck = false;
    input.value = item.line;
    row.appendChild(input);
    if (['┼', '┬'].includes(edit.textContent)) {
        edit.innerHTML += `<br>│`.repeat(pre.offsetHeight / 15 - 1);
    }
    edit.addEventListener('click', event => {
        edit.style.display = 'none';
        pre.style.display = 'none';
        none.style.display = 'block';
        input.style.display = 'block';
        input.focus();
        const length = input.value.length;
        input.setSelectionRange(length, length);
        row.style.background = 'aliceblue';
    });
    input.addEventListener('keypress', async event => {
        if (event.key !== 'Enter') { return; }
        const item_ = { id: item.id, line: item.line, revision: input.value };
        if (!postEdit(item_)) { alert('Error!'); }
        location.reload();
    });
};

const download = event => {
    const element = document.createElement('a');
    const data = encodeURIComponent(localStorage.clutter);
    element.setAttribute('href', 'data:text/plain;charset=utf-8,' + data);
    element.setAttribute('download', header().trim() + '.txt');
    element.click();
};

const upload = event => {
    const reader = new FileReader();
    const element = document.createElement('input');
    element.type = 'file';
    element.addEventListener('change', event => {
        reader.addEventListener('load', event => {
            localStorage.clutter = event.target.result;
            location.href = location.pathname;
        });
        reader.readAsText(event.target.files[0]);
    });
    element.click();
};

const onIndex = data => {
    document.getElementById('getset').style.display = 'flex';
    document.getElementById('new').style.display = 'flex';
    document.getElementById('indexRows').style.display = 'block';
    const get = document.getElementById('get');
    get.addEventListener('click', download );
    const set = document.getElementById('set');
    set.addEventListener('click', upload );
    const add = document.getElementById('add');
    add.addEventListener('keypress', async event => {
        if (event.key !== 'Enter') { return; }
        push(add.value);
        location.reload();
    });
    const plus = document.getElementById('plus');
    plus.addEventListener('click', event => {
        add.value = search.value;
        add.focus();
    });
    for (const item of data.items.slice(0, 1000)) {
        createIndex(item);
    }
    if (data.items.length > 1000) {
        document.getElementById('overflow').style.display = 'flex';
    }
};

const main = () => {
    const params = new URLSearchParams(location.search);
    if (params.has('add')) {
        push(params.get('add'));
        history.replaceState({}, '', location.pathname);
    }
    if (performance.getEntriesByType('navigation')[0].type === 'back_forward') {
        location.reload();
    }
    const q = params.get('q') || '';
    const query = parse(q);
    const data = process(query);
    document.title = data.title;
    document.getElementById('count').innerHTML = data.count;
    const search = document.getElementById('search');
    search.value = data.q;
    search.addEventListener('keypress', event => {
        if (event.key !== 'Enter') { return; }
        const q = encodeURIComponent(search.value);
        location.href = '?q=' + q;
    });
    if (query.swap) {
        onSwap(data);
    }
    else {
        onIndex(data);
    }
    document.querySelectorAll('.left').forEach(element => {
        element.style.visibility = 'visible';
    });
};

const postEdit = item => {
    const [lines, n] = read();
    const index = n - item.id;
    if (lines[index] !== item.line) {
        return false;
    }
    lines[index] = item.revision;
    write(lines);
    return true;
};

const postSwap = items => {
    const [lines, n] = read();
    for (const item of items) {
        const index = n - item.id;
        if (lines[index] !== item.line) {
            return false;
        }
        lines[index] = item.revision;
        write(lines);
    }
    return true;
};

const link = line => {
    const parts = [];
    tag = { '#': 'hash', '$': 'dollar', '^': 'caret' };
    for (const word of line.split(' ')) {
        if (word.startsWith('http')) {
            parts.push(`<a href="${escape(word)}">${escape(word)}</a>`);
        }
        else if (word && word[0] in tag) {
            parts.push(`<span class="${tag[word[0]]}">${escape(word)}</span>`);
        }
        else {
            parts.push(escape(word));
        }
    }
    line = parts.join(' ');
    if (line.match(/\d{14}/)) {
        const year = `<span class="timestamp">${line.slice(0, 4)}</span>`;
        const a = `<a href="?q=${line.slice(0, 8)}">${line.slice(4, 8)}</a>`;
        const moment = `<span class="timestamp">${line.slice(8, 14)}</span>`;
        const body = line.slice(14);
        line = `${year}${a}${moment}${body}`;
    }
    return line;
};

main();
